File: bindings/typescript/tests/client.test.ts
/**
 * Binding tests: create client, request, deserialize, handle error, async.
 * Fully offline: runs against a local Bun.serve mock.
 */

import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import {
  ApiError,
  HttpError,
  InvalidParameterError,
  JourneyPlace,
  RateLimitedError,
  SerializationError,
  CapabilityNotSupportedError,
  TransportRestClient,
} from "../src/index.js";


const WHEN = new Date("2026-08-01T12:00:00+02:00");

let server: ReturnType<typeof Bun.serve>;
let baseUrl = "";

beforeAll(() => {
  server = Bun.serve({
    port: 0,
    fetch(request) {
      const url = new URL(request.url);
      const q = Object.fromEntries(url.searchParams.entries());
      const json = (status: number, body: unknown, headers?: Record<string, string>) =>
        new Response(typeof body === "string" ? body : JSON.stringify(body), {
          status,
          headers: { "Content-Type": "application/json", ...headers },
        });
      switch (`${url.pathname}`) {
        case "/locations":
          if (q.query === "Berlin") {
            return json(200, [
              { type: "stop", id: "8011160", name: "Berlin Hbf",
                location: { type: "location", latitude: 52.525, longitude: 13.369 } },
              { type: "location", name: "Alexanderplatz", poi: true },
            ]);
          }
          if (q.query === "ratelimit") {
            return json(429, { message: "Too Many Requests" }, { "Retry-After": "30" });
          }
          if (q.query === "badgateway") {
            return new Response("<html>Bad Gateway</html>", { status: 502 });
          }
          return json(400, { message: "Missing query." });
        case "/stops/8011160/departures":
          return json(200, {
            departures: [{
              tripId: "t1",
              line: { id: "ICE 599", mode: "train" },
              when: "2026-08-01T12:00:00+02:00",
              plannedWhen: "2026-08-01T11:58:00+02:00",
              delay: 120,
            }],
            realtimeDataUpdatedAt: 1754000000,
          });
        case "/journeys":
          return json(200, {
            journeys: [{
              refreshToken: "ref/1",
              legs: [{
                tripId: "t9",
                origin: { type: "stop", id: "8011160", name: "Berlin Hbf" },
                destination: { type: "stop", id: "8000108", name: "Leipzig Hbf" },
                line: { id: "ice-599", mode: "train",
                        operator: { type: "operator", id: "db", name: "Deutsche Bahn" } },
              }],
            }],
            earlierRef: "E",
            laterRef: "L",
          });
        case "/stops/nope":
          return json(404, { message: "Stop not found." });
        case "/radar":
          return json(200, { movements: [{ tripId: "m1" }] });
        case "/stops/x/departures":
          return json(200, "{not json");
        default:
          return json(404, { message: `no route ${url.pathname}` });
      }
    },
  });
  baseUrl = `http://localhost:${server.port}`;
});

afterAll(() => server.stop(true));

const dbClient = () => new TransportRestClient({ provider: "db", baseUrl });

describe("locations", () => {
  test("parses results and sends query params", async () => {
    const client = dbClient();
    const result = await client.locations().query("Berlin").results(5).get();
    expect(result).toHaveLength(2);
    expect((result[0] as { id?: string }).id).toBe("8011160");
  });

  test("missing query fails client-side without request", async () => {
    const client = dbClient();
    await expect(client.locations().get()).rejects.toBeInstanceOf(InvalidParameterError);
  });

  test("429 exposes retry-after", async () => {
    const client = dbClient();
    try {
      await client.locations().query("ratelimit").get();
      throw new Error("expected RateLimitedError");
    } catch (err) {
      expect(err).toBeInstanceOf(RateLimitedError);
      expect((err as RateLimitedError).retryAfterSeconds).toBe(30);
    }
  });

  test("non-JSON 502 becomes HttpError", async () => {
    const client = dbClient();
    try {
      await client.locations().query("badgateway").get();
      throw new Error("expected HttpError");
    } catch (err) {
      expect(err).toBeInstanceOf(HttpError);
      expect((err as HttpError).status).toBe(502);
    }
  });
});

describe("departures", () => {
  test("parses board and product filters", async () => {
    const client = dbClient();
    const result = await client
      .departures("8011160")
      .results(10)
      .products((p) => p.bus(false).tram(false))
      .moreStops(["8010159"])
      .when(WHEN)
      .get() as any;
    expect(result.departures[0].delay).toBe(120);
    expect(result.departures[0].line.mode).toBe("train");
    expect(result.realtimeDataUpdatedAt).toBe(1754000000);
  });
});

describe("journeys", () => {
  test("encodes places and pagination refs", async () => {
    const client = dbClient();
    const result = await client
      .journeys(JourneyPlace.stopId("8011160"), JourneyPlace.name("Leipzig Hbf"))
      .via(JourneyPlace.poi("poi1", 51.5, 12.2))
      .transfers(0)
      .earlierThan("REF1")
      .get();
    const journey = result.journeys?.[0];
    expect(journey?.legs?.[0]?.origin).toMatchObject({ id: "8011160" });
    expect(journey?.refreshToken).toBe("ref/1");
  });

  test("conflicting times rejected before any request", async () => {
    const client = dbClient();
    await expect(
      client.journeys(JourneyPlace.stopId("a"), JourneyPlace.stopId("b"))
        .departure(WHEN).arrival(WHEN).get(),
    ).rejects.toBeInstanceOf(InvalidParameterError);
  });
});

describe("errors & capabilities", () => {
  test("structured API error keeps status and message", async () => {
    const client = dbClient();
    try {
      await client.stop("nope").get();
      throw new Error("expected ApiError");
    } catch (err) {
      expect(err).toBeInstanceOf(ApiError);
      expect((err as ApiError).status).toBe(404);
      expect((err as ApiError).message).toBe("Stop not found.");
    }
  });

  test("capability gating: radar on db vs bvg", async () => {
    const db = dbClient();
    await expect(
      db.radar().north(52.53).west(13.36).south(52.51).east(13.39).get(),
    ).rejects.toBeInstanceOf(CapabilityNotSupportedError);

    const bvg = new TransportRestClient({ provider: "bvg", baseUrl });
    const radar = await bvg.radar()
      .north(52.53).west(13.36).south(52.51).east(13.39).get();
    expect(radar.movements?.[0]?.tripId).toBe("m1");
  });

  test("malformed JSON is a serialization error", async () => {
    const client = dbClient();
    await expect(client.departures("x").get())
      .rejects.toBeInstanceOf(SerializationError);
  });

  test("unknown future fields are tolerated", async () => {
    const client = dbClient();
    // /stops/nope returns 404; use journeys response with extra field instead:
    const result = await client
      .journeys(JourneyPlace.stopId("8011160"), JourneyPlace.name("L"))
      .get();
    expect(Object.keys(result)).toContain("earlierRef");
  });
});

File: bindings/go/client_test.go
// Binding tests: create client, request, deserialize, handle error.
// Fully offline via net/http/httptest.

package transportrest

import (
	"context"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func newTestClient(t *testing.T, serverURL string) *Client {
	t.Helper()
	c, err := NewClient(ClientOptions{Provider: "db", BaseURL: serverURL})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return c
}

func TestLocationsHappyPath(t *testing.T) {
	var gotPath, gotQuery string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		gotQuery = r.URL.RawQuery
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`[
			{"type":"stop","id":"8011160","name":"Berlin Hbf",
			 "location":{"type":"location","latitude":52.525,"longitude":13.369}},
			{"type":"location","name":"Alexanderplatz","poi":true}
		]`))
	}))
	defer srv.Close()

	client := newTestClient(t, srv.URL)
	result, err := client.Locations(context.Background()).Query("Berlin").Results(5).Get()
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if gotPath != "/locations" || !strings.Contains(gotQuery, "query=Berlin") {
		t.Fatalf("unexpected request: %s?%s", gotPath, gotQuery)
	}
	if len(result) != 2 {
		t.Fatalf("expected 2 results, got %d", len(result))
	}
	if result[0]["id"] != "8011160" {
		t.Fatalf("stop id mismatch: %+v", result[0])
	}
}

func TestLocationsRequiresQuery(t *testing.T) {
	client := newTestClient(t, "http://127.0.0.1:1")
	_, err := client.Locations(context.Background()).Get()
	if err == nil {
		t.Fatal("expected InvalidParameterError")
	}
	if _, ok := err.(*InvalidParameterError); !ok {
		t.Fatalf("wrong error type: %T", err)
	}
}

func TestDeparturesParsesBoard(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !strings.HasSuffix(r.URL.Path, "/departures") {
			t.Errorf("path: %s", r.URL.Path)
		}
		if q := r.URL.Query().Get("bus"); q != "false" {
			t.Errorf("product filter missing: %q", q)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"departures":[{"tripId":"t1",
			"line":{"id":"ICE 599","mode":"train"},"delay":120.0,
			"when":"2026-08-01T12:00:00+02:00"}],
			"realtimeDataUpdatedAt":1754000000}`))
	}))
	defer srv.Close()

	client := newTestClient(t, srv.URL)
	result, err := client.Departures(context.Background(), "8011160").
		Products(func(p *ProductSelection) *ProductSelection { return p.Bus(false).Tram(false) }).
		Get()
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if result.RealtimeDataUpdatedAt == nil || *result.RealtimeDataUpdatedAt != 1754000000 {
		t.Fatalf("realtime timestamp mismatch")
	}
	if len(result.Departures) != 1 {
		t.Fatalf("departures: %d", len(result.Departures))
	}
}

func TestRateLimitedWithRetryAfter(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Retry-After", "30")
		w.WriteHeader(http.StatusTooManyRequests)
		_, _ = w.Write([]byte(`{"message":"Too Many Requests"}`))
	}))
	defer srv.Close()

	client := newTestClient(t, srv.URL)
	_, err := client.Locations(context.Background()).Query("x").Get()
	rl, ok := err.(*RateLimitedError)
	if !ok {
		t.Fatalf("expected RateLimitedError, got %T (%v)", err, err)
	}
	if !rl.HasRetryAfter() || rl.RetryAfter != 30*time.Second {
		t.Fatalf("retry-after: %+v", rl)
	}
}

func TestStructuredApiError404(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"message":"Stop not found."}`))
	}))
	defer srv.Close()

	client := newTestClient(t, srv.URL)
	_, err := client.Locations(context.Background()).Query("x").Get()
	api, ok := err.(*ApiError)
	if !ok {
		t.Fatalf("expected ApiError, got %T", err)
	}
	if api.Status != 404 || api.Message != "Stop not found." {
		t.Fatalf("api error: %+v", api)
	}
}

func TestNonJson502BecomesHttpError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusBadGateway)
		_, _ = w.Write([]byte("<html>Bad Gateway</html>"))
	}))
	defer srv.Close()

	client := newTestClient(t, srv.URL)
	_, err := client.Locations(context.Background()).Query("x").Get()
	herr, ok := err.(*HttpError)
	if !ok {
		t.Fatalf("expected HttpError, got %T", err)
	}
	if herr.Status != 502 || !strings.Contains(herr.BodySnippet, "<html>") {
		t.Fatalf("http error: %+v", herr)
	}
}

func TestCapabilityGating(t *testing.T) {
	client := newTestClient(t, "http://127.0.0.1:1")
	if client.Supports("radar") {
		t.Fatal("db must not support radar by default")
	}
	custom, err := NewClient(ClientOptions{
		BaseURL: "http://127.0.0.1:1", Provider: "custom", EnableCapabilities: []string{"radar"},
	})
	if err != nil {
		t.Fatalf("custom client: %v", err)
	}
	if !custom.Supports("radar") {
		t.Fatal("enabled capability missing")
	}
}

func TestUnknownFieldsAreIgnored(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"type":"stop","id":"futuristic","brandNewField":[1,2]}`))
	}))
	defer srv.Close()

	client := newTestClient(t, srv.URL)
	stop, err := client.Locations(context.Background()).Query("f").Get()
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if len(stop) != 1 || stop[0]["type"] != "stop" {
		t.Fatalf("unexpected: %+v", stop)
	}
}
